src/math/exp_fcn.c

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+/* Include Files */
+#include <sof/math/exp_fcn.h>
+#include <sof/math/numbers.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+//inv multiplication lookup table
+// LUT = ceil(1/factorial(n) * 2 ^ 63)
+static const int64_t iv[19] = {
+		4611686018427387904LL,
+		1537228672809129301LL,
+		384307168202282325LL,
+		76861433640456465LL,
+		12810238940076077LL,
+		1830034134296582LL,
+		228754266787072LL,
+		25417140754119LL,
+		2541714075411LL,
+		231064915946LL,
+		19255409662LL,
+		1481185358LL,
+		105798954LL,
+		7053264LL,
+		440829LL,
+		25931LL,
+		1441LL,
+		76LL,
+		4LL
+};
+
+/* Function Declarations */
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo);
+/*
+ * Arguments	: int64_t a
+ *		  int64_t b
+ * Return Type	: int64_t
+ */
+static inline int64_t lomul_s64_SR_sat_near(int64_t a, int64_t b)
+{
+	uint64_t u64_rhi;
+	uint64_t u64_rlo;
+	bool roundup;
+
+	wide_mul_s64(a, b, &u64_rhi, &u64_rlo);
+	roundup = (u64_rlo & BIT_MASK_LOW_Q27P5);
+	u64_rlo = (u64_rhi << 36 | u64_rlo >> 28) + roundup;
+	return u64_rlo;
+}
+
+/*
+ * Arguments	: int64_t in_0
+ *		  int64_t in_1
+ *		  uint64_t *ptroutbitshi
+ *		  uint64_t *ptroutbitslo
+ * Multiplication of two signed int64 numbers
+ * Return Type	: void
+ */
+static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
+				uint64_t *ptroutbitslo)
```

Comment:
What's the value to having this in code here vs. just relying on the toolchain?  My experience is that most platform runtimes are pretty optimal, I have a very hard time imagining that we're going to beat them in portable C (and for sure this is going to bloat code size, since something else is already doing a 64 bit multiplication anyway.

Basically:  if we replace this with the following code, what would be worse:

```c
static inline void wide_mul_s64(int64_t in_0, int64_t in_1, uint64_t *ptroutbitshi,
				uint64_t *ptroutbitslo)
{
    *ptroutbitslo = in_0 * in_1;
    *ptroutbitshi = *ptroutbitslo >> 32;
    *ptroutbitslo &= 0xffffffffull;
}
```

Reply:
I don't understand the response?  The toolchain's built-in 64 bit integer multiplication routines absolutely don't lose accuracy, they implement 2's complement math exactly.  Is there something this function does that a standard 64 bit multiplication doesn't?  I don't see it if so.

Reply:
I'm sending some sample data with a working case (using the current code) and a broken case for reference ( your suggestion ).


Both situations have the same input, which is -5, and the same expected result, which is "exp(-5)" 0.0067. ( 4 fractional place of accuracy) 

Thank you !!

```
working case
               absIn0|                absIn1|                 in0Hi|                 in0Lo|                 in1Hi|                  in_0|                  in_1|          ptroutbitshi|          ptroutbitslo|
---------------------+----------------------+----------------------+----------------------+----------------------+----------------------+----------------------+----------------------+----------------------|+
 18446744073709531136|                     0|                     0|                     0|            1073741824|                -81920|   4611686018427387904|  18446744073709531136|                     0|
                    0|                     0|                     0|       109951162777600|                     0|                -81920|           -1342177280|                     0|       109951162777600|
               102400|                     0|                     0|                     0|            1073741824|                409600|   4611686018427387904|                102400|                     0|
 18446744073709551615|                     0|                     0|  18446194317895663616|                     0|                409600|           -1342177280|  18446744073709551615|  18446194317895663616|
 18446744073709380949|                     0|                     0|   6148914691237199872|             357913941|              -2048000|   1537228672809129301|  18446744073709380949|   6148914691237199872|
                    0|                     0|                     0|      2748779069440000|                     0|              -2048000|           -1342177280|                     0|      2748779069440000|
               213333|                     0|                     0|   6148914691233103872|              89478485|              10240000|    384307168202282325|                213333|   6148914691233103872|
 18446744073709551615|                     0|                     0|  18433000178362351616|                     0|              10240000|           -1342177280|  18446744073709551615|  18433000178362351616|
 18446744073709338282|                     0|                     0|  12297829382476447744|              17895697|             -51200000|     76861433640456465|  18446744073709338282|  12297829382476447744|
                    0|                     0|                     0|     68719476736000000|                     0|             -51200000|           -1342177280|                     0|     68719476736000000|
               177777|                     0|                     0|  14347467612754362368|               2982616|             256000000|     12810238940076077|                177777|  14347467612754362368|
 18446744073709551615|                     0|                     0|  18103146690029551616|                     0|             256000000|           -1342177280|  18446744073709551615|  18103146690029551616|
 18446744073709424631|                     0|                     0|  16104300382451957760|                426088|           -1280000000|      1830034134296582|  18446744073709424631|  16104300382451957760|
                    0|                     0|                     0|   1717986918400000000|                     0|           -1280000000|           -1342177280|                     0|   1717986918400000000|
                79365|                     0|                     1|   1464027302235996160|                 53261|            6400000000|       228754266787072|                 79365|   1464027302235996160|
 18446744073709551615|                     0|                     1|   9856809481709551616|                     0|            6400000000|           -1342177280|  18446744073709551615|   9856809481709551616|
 18446744073709507524|                     0|                     7|   5335566193549852672|                  5917|          -32000000000|        25417140754119|  18446744073709507524|   5335566193549852672|
                    2|                     0|                     7|   6056184812580896768|                     0|          -32000000000|           -1342177280|                     2|   6056184812580896768|
                22045|                     0|                    37|  15778960832934625280|                   591|          160000000000|         2541714075411|                 22045|  15778960832934625280|
 18446744073709551604|                     0|                    37|   6612564084514619392|                     0|          160000000000|           -1342177280|  18446744073709551604|   6612564084514619392|
 18446744073709541595|                     0|                   186|   2889605843416743936|                    53|         -800000000000|          231064915946|  18446744073709541595|   2889605843416743936|
                   58|                     0|                   186|   3830667724846006272|                     0|         -800000000000|           -1342177280|                    58|   3830667724846006272|
                 4175|                     1|                   931|   6482140262622003200|                     4|         4000000000000|           19255409662|                  4175|   6482140262622003200|
 18446744073709551324|                     1|                   931|  17740149523189071872|                     0|         4000000000000|           -1342177280|  18446744073709551324|  17740149523189071872|
 18446744073709550010|                     0|                  4656|   1763822377539895296|                     0|       -20000000000000|            1481185358|  18446744073709550010|   1763822377539895296|
                 1455|                     0|                  4656|   3532972752602398720|                     0|       -20000000000000|           -1342177280|                  1455|   3532972752602398720|
                  573|                     0|                 23283|   9911045764426924032|                     0|       100000000000000|             105798954|                   573|   9911045764426924032|
 18446744073709544340|                     0|                 23283|    781880310697558016|                     0|       100000000000000|           -1342177280|  18446744073709544340|    781880310697558016|
 18446744073709551424|                     0|                116415|  15142862152233910272|                     0|      -500000000000000|               7053264|  18446744073709551424|  15142862152233910272|
                36379|                     0|                116415|  14537342520221761536|                     0|      -500000000000000|           -1342177280|                 36379|  14537342520221761536|
                   59|                     0|                582076|  13714599651136454656|                     0|      2500000000000000|                440829|                    59|  13714599651136454656|
 18446744073709369717|                     0|                582076|   1100263693729398784|                     0|      2500000000000000|           -1342177280|  18446744073709369717|   1100263693729398784|
 18446744073709551598|                     0|               2910383|   7903893326771929088|                     0|    -12500000000000000|                 25931|  18446744073709551598|   7903893326771929088|
               909494|                     0|               2910383|  12945425605062557696|                     0|    -12500000000000000|           -1342177280|                909494|  12945425605062557696|
                    4|                     0|              14551915|  16275523705161793536|                     0|     62500000000000000|                  1441|                     4|  16275523705161793536|
 18446744073705004142|                     0|              14551915|   9059848269525417984|                     0|     62500000000000000|           -1342177280|  18446744073705004142|   9059848269525417984|
 18446744073709551614|                     0|              72759576|  13143488147419103232|                     0|   -312500000000000000|                    76|  18446744073709551614|  13143488147419103232|
             22737367|                     0|              72759576|  10040990873501564928|                     0|   -312500000000000000|           -1342177280|              22737367|  10040990873501564928|
                    0|                     0|             363797880|   6250000000000000000|                     0|   1562500000000000000|                     4|                     0|   6250000000000000000|
 18446744073595864778|                     0|             363797880|   5135277853620830208|                     0|   1562500000000000000|           -1342177280|  18446744073595864778|   5135277853620830208|
 
   input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN|
---------------+---------------+---------------+---------------+
    -5.00000000|     0.00674736|     0.00673795|  -100.52082740|
    
    

broken case

                 in_0|                  in_1|          ptroutbitshi|          ptroutbitslo|
---------------------+----------------------+----------------------+----------------------+		 
               -81920|   4611686018427387904|                     0|                     0|
               -81920|           -1342177280|                 25600|                     0|
     1759218604441600|   4611686018427387904|                     0|                     0|
     1759218604441600|           -1342177280|                     0|                     0|
     
   input(Q4.28)|    r_hw(Q9.23)|         r_gold|           THDN| 
---------------+---------------+---------------+---------------+
    -5.00000000|     1.00000000|     0.00673795|    -0.05872312|

```

Reply:
Sorry, I missed the update and @marc-hb had to poke me about this.

I... actually I think that pushes me into -1 territory here.  The name "wide_mul_s64" clearly implies that this function is computing a 64 big signed product of its two arguments.  But your test shows that... it's not?!  Then what exactly is it doing?

Can you update this to explain exactly how this differs from a 2's complement signed multiply?  Otherwise it's just a booby trap waiting for someone to come by and "fix" it and produce the precision errors you report.

(Alternatively and more likely, one of us just messed up the sample code and this is just a perfectly normal multiply.  It definitely looks like it to me.)

Reply:
@ShriramShastry I think what you need to do here, is to explain in the function header comments why this function is different, better and faster than a 64bit vanilla multiple AND where and when it should and should not be used. This way its easier for non expert reviewers and users to understand the usage.

Reply:
Sure